shop/src/main/java/ro/msg/learning/shop/services/ProductService.java

```diff
@@ -0,0 +1,102 @@
+package ro.msg.learning.shop.services;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import ro.msg.learning.shop.dtos.ProductCategoryDTO;
+import ro.msg.learning.shop.dtos.ProductDTO;
+import ro.msg.learning.shop.entities.Product;
+import ro.msg.learning.shop.entities.ProductCategory;
+import ro.msg.learning.shop.exceptions.ProductNotFoundException;
+import ro.msg.learning.shop.mappers.ProductMapper;
+import ro.msg.learning.shop.repositories.ProductCategoryRepository;
+import ro.msg.learning.shop.repositories.ProductRepository;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class ProductService {
+    private final ProductRepository productRepository;
+    private final ProductMapper productMapper;
+    private final ProductCategoryRepository productCategoryRepository;
+
+    public ProductDTO createProduct(ProductDTO productDto) {
+        Product product = Product.builder()
+                .name(productDto.getName())
+                .description(productDto.getDescription())
+                .price(productDto.getPrice())
+                .weight(productDto.getWeight())
+                .imageUrl(productDto.getImageUrl())
+                .productCategory(isCategory(productDto.getProductCategory()))
+                .build();
+        productRepository.save(product);
+        return productMapper.productToProductDTO(product);
+    }
+
+    public ProductCategory isCategory(ProductCategoryDTO productCategoryDTO) {
+        Optional<ProductCategory> searchedCategory = productCategoryRepository.findByName(productCategoryDTO.getName());
+        ProductCategory productCategory;
+        if (searchedCategory.isPresent()) {
+            productCategory = searchedCategory.get();
+        } else {
+            productCategory = new ProductCategory();
+            productCategory.setName(productCategoryDTO.getName());
+            productCategory.setDescription(productCategoryDTO.getDescription());
+            productCategoryRepository.save(productCategory);
+        }
+        return productCategory;
+
+    }
+
+    public ProductDTO updateProduct(Integer id, ProductDTO updatedProduct) {
+        ProductDTO resultedProduct = null;
```

Comment:
if there is no resulted product, the returned value will be NULL. This is not good practice. When a product does not exist it should be handled with an exception and a message. No problem can be solved forever by giving back a null, but many other problems can be created 😄